src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperDescriptorFactory.cs

```diff
@@ -62,11 +73,105 @@ private static IEnumerable<string> GetTagNames(Type tagHelperType)
-            return attributes.SelectMany(attribute => attribute.Tags).Distinct();
+            return targetElementAttributes.SelectMany(
+                targetElementAttribute =>
+                {
+                    IEnumerable<string> tagNames;
+                    IEnumerable<string> attributeNames = null;
+
+                    if (!TryGetValidatedNames(
+                            targetElementAttribute.Tags,
+                            pascalNameTarget: "Tag",
+                            errorSink: errorSink,
+                            names: out tagNames) ||
+                        (targetElementAttribute.Attributes != null &&
+                        !TryGetValidatedNames(
+                            targetElementAttribute.Attributes,
+                            pascalNameTarget: "Attribute",
+                            errorSink: errorSink,
+                            names: out attributeNames)))
+                    {
+                        return Enumerable.Empty<ElementTarget>();
+                    }
+
+                    return BuildElementTargets(tagNames, attributeNames, tagHelperType);
+                }).ToArray();
+        }
+
+        // Internal for testing
+        internal static bool TryGetValidatedNames(
+            string commaSeparatedNames,
+            string pascalNameTarget,
+            ParserErrorSink errorSink,
+            out IEnumerable<string> names)
+        {
+            names = GetCommaSeparatedValues(commaSeparatedNames)?.Distinct();
+
+            return ValidateNames(names, pascalNameTarget, errorSink);
+        }
+
+        private static IEnumerable<string> GetCommaSeparatedValues(string text)
+        {
+            // We don't want to remove empty entries, need to notify users of invalid values.
+            return text.Split(',').Select(tagName => tagName.Trim());
+        }
+
+        private static bool ValidateNames(
+            IEnumerable<string> names,
+            string pascalNameTarget,
+            ParserErrorSink errorSink)
+        {
+            foreach (var name in names)
+            {
+                if (string.IsNullOrWhiteSpace(name))
+                {
+                    errorSink.OnError(
+                        SourceLocation.Zero,
+                        Resources.FormatTargetElementAttribute_NameCannotBeNullOrWhitespace(pascalNameTarget));
+
+                    return false;
+                }
+
+                foreach (var character in name)
+                {
+                    if (InvalidNonWhitespaceNameCharacters.Contains(character))
+                    {
+                        errorSink.OnError(
+                            SourceLocation.Zero,
+                            Resources.FormatTargetElementAttribute_InvalidName(
+                                pascalNameTarget.ToLowerInvariant(),
+                                name,
+                                character));
+
+                        return false;
```

Comment:
these early exits mean the method will only add an error for the first invalid name and (for non-whitespace issues) the first invalid character in that invalid name.  better to inform the user of _all_ problems in all names.  that way they can fix all problems at once.